scheduler/reconcile/reconcile_cluster.go

```diff
@@ -237,7 +227,7 @@ func NewAllocReconciler(logger log.Logger, allocUpdateFn allocUpdateType, batch
 
 // Compute reconciles the existing cluster state and returns the set of changes
 // required to converge the job spec and state
-func (a *allocReconciler) Compute() *reconcileResults {
```

Comment:
I have a slight preference for keeping the `result` field private and returning it from the `Compute()` method as (1) if you only use methods (and not struct fields directly), you can create a reconciler interface (which might be useful for testing/debugging), and (2) exported methods with only-side-effects like `Node.ComputeClass()` make it really easy to create bugs. It's not obvious to callers _when_ they should be called. If it's immediately after `NewX()` ... when not bake it into `NewX()`? There's often a reason (eg you have to poke at fields a bit and _then_ call the side-effect method), but it's rarely clear and makes bugs easy.

Reply:
@pkazmierczak and I had a discussion about that earlier this morning and this caught my eye too. We're all in agreement that we want to reduce/eliminate all the hard-to-reason-about mutation all over the scheduler code base. But Piotr argued convincingly for me that, for the moment, having this method signature _reveal_ that it's full of stinky side-effects is helpful to make it obvious that there's a next step of work to do. There are going to be a bunch more PRs as we go here.